Add SSE connector

hseeberger opened the pull request:
Java API and docs missing and depending on a milestone of akka-sse. But the Scala API and tests are ready for review.

ktoso commented:
I'm not sure about it in Alpakka, why not move it closer to Akka HTTP instead?

hseeberger commented:
@ktoso I would like to move akka-sse itself into Akka HTTP, but this is a connector to an SSE server.

ktoso commented:
I see. Yes that makes sense.
Will review in a bit then :)

On Nov 26, 2016 17:07, "Heiko Seeberger" <notifications@github.com> wrote:

> @ktoso <https://github.com/ktoso> I would like to move akka-sse itself
> into Akka HTTP, but this is a connector to an SSE server.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/akka/alpakka/pull/88#issuecomment-263071254>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHYk0jfNxtETKDgGQFHXlcfdp41XztNks5rCFk9gaJpZM4K86Eh>
> .
>

hseeberger commented:
@viktorklang You have already reviewed the now deprecated `EventStreamClient` in akka-sse – this one is very similar, but instead of handling (by means of a sinnk) the event streams from the server, the `EventSource` "just" represents a continuous reconnecting source.

viktorklang commented:
@hseeberger I don't know what you want me to do here :)

hseeberger commented:
@viktorklang Well, usually I love your reviews and as you have already reviewed this in slightly modified form I was hoping I would like to take another look ...

viktorklang commented:
@hseeberger Ah, ok, I'll have a look. I just opened a PR against akka-sse ;)

viktorklang commented on sse/src/main/scala/akka/stream/alpakka/sse/EventSource.scala:
```diff
@@ -0,0 +1,126 @@
+      def enrichWithLastElement(eventStream: EventStream) = {
+        val p = Promise[LastEvent]()
+        val enriched =
+          eventStream.alsoToMat(Sink.lastOption)(Keep.both).mapMaterializedValue {
```
See: https://github.com/hseeberger/akka-sse/pull/158/files

viktorklang commented on sse/src/main/scala/akka/stream/alpakka/sse/EventSource.scala:
```diff
@@ -0,0 +1,126 @@
+          }
+        (enriched, p.future)
+      }
+      Flow[Option[String]].mapAsync(1)(getEventStream).map(enrichWithLastElement)
```
Is the `1` here crucial?

hseeberger replied:
There can only ever be one request in flight, so I think that `1` is a nice fit. Any other higher number wouldn't change anything.

viktorklang replied:
Good point, but might make sense to document somewhere in the code :)

hseeberger replied:
Right.

viktorklang commented on sse/src/test/scala/akka/stream/alpakka/sse/EventSourceSpec.scala:
```diff
@@ -0,0 +1,156 @@
+      get {
+        optionalHeaderValueByName(`Last-Event-ID`.name) { lastEventId =>
+          try {
+            val fromSeqNo = lastEventId.getOrElse("0").trim.toInt + 1
```
Overflow protection?

viktorklang replied:
`lastEventId.map(_.trim.toInt).getOrElse(0) + 1` ?

hseeberger replied:
Is this to avoid calling `trim.toInt` on "0"?

viktorklang replied:
Yes, it's more typed and stuff.

hseeberger replied:
+1, yet almost pedantic for test code ;-)

viktorklang commented on sse/src/test/scala/akka/stream/alpakka/sse/EventSourceSpec.scala:
```diff
@@ -0,0 +1,156 @@
+          try {
+            val fromSeqNo = lastEventId.getOrElse("0").trim.toInt + 1
+            complete {
+              Source.fromIterator(() => Iterator.from(fromSeqNo).take(2)).map(toServerSentEvent(setEventId))
```
Isn't that `Source(fromSeqNo to (fromSeqNo + 2))` ?

hseeberger replied:
Same semantics, yes. Any advantages?

viktorklang replied:
Besides being shorter I think it also keeps more semantic information :)

hseeberger replied:
You're the boss ;-)

viktorklang commented on sse/src/test/scala/akka/stream/alpakka/sse/EventSourceSpec.scala:
```diff
@@ -0,0 +1,156 @@
+  private def toServerSentEvent(setEventId: Boolean)(n: Int) = {
+    val id = n.toString
+    val event = ServerSentEvent(Some(id))
+    if (setEventId) event.copy(id = Some(id)) else event
```
Why allocate `Some(id)` twice?

viktorklang replied:
Also, the code looks weird. It creaes a new SSE with the id, then copies it with the same id?

hseeberger replied:
It's a test, mate ;-)

hseeberger replied:
Would it help if I rewrote that to `val event = ServerSentEvent(data = Some(id))`?

hseeberger replied:
I rewrite the whole section such that it makes more sense when reading ;-)

viktorklang commented on sse/src/test/scala/akka/stream/alpakka/sse/EventSourceSpec.scala:
```diff
@@ -0,0 +1,156 @@
+  }
+
+  override protected def afterAll() = {
+    Await.ready(system.terminate(), 42.seconds)
```
Arbitrary number?

hseeberger replied:
42 is not arbitrary! ;-)
Back in the early days I used `Duration.Inf` until @patriknw (I think it was some ConductR code) told me that it's better to use some finite value. He was fine with `42`.

viktorklang replied:
Ok ;)

patriknw replied:
yeah, at least it doesn't make the CI build hang forever
(a realistic value is 10.seconds, but I'm still fine with 42)

hseeberger commented on sse/src/test/scala/akka/stream/alpakka/sse/EventSourceSpec.scala:
```diff
@@ -0,0 +1,156 @@
+  "EventSource" should {
+    "communicate correctly with an instable HTTP server" in {
+      val host = "localhost"
+      val port = 9999
```
Agreed. How would I easily obtain one?

johanandren replied:
I think `akka.stream.testkit.TestUtils#temporaryServerAddress` is public

hseeberger replied:
@johanandren I can't find that, but there's `akka.util.SocketUtil#temporaryServerAddress`.

johanandren replied:
Yeah, that would need dep on the stream testkit I guess, the akka util one should be just as good (not quite sure why we have two).

viktorklang commented on sse/src/test/scala/akka/stream/alpakka/sse/EventSourceSpec.scala:
```diff
@@ -0,0 +1,159 @@
+  private def send(request: HttpRequest) = Http().singleRequest(request)
+
+  private def hostAndPort() = {
+    val address = SocketUtil.temporaryServerAddress()
```
:+1:

hseeberger commented:
I have addressed all points from @viktorklang. In order to procees, which involves removing `EventSource` from akka-sse and releasing akka-sse 2.0.0 final, I need to know if there are any general objections against adding this to Alpakka.

patriknw commented:
@hseeberger haven't reviewed in detail but looking good and I have no objections with including in Alpakka

drewhk commented:
LGTM, but needs a rebase.

patriknw commented:
@hseeberger what's the status here?

hseeberger commented:
Rebased. On top of Akka SSE 2.0. Ready for review except for Java API and docs (ScalaDoc does exist).

What about package names of "Scala DSL" and "Akka DSL" (these are no DSLs, are they)?

hseeberger commented:
Akka team, please review. Thanks!

patriknw commented on sse/src/test/scala/akka/stream/alpakka/sse/EventSourceSpec.scala:
```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.sse
```
Should this be in `scaladsl` to follow the convention of other things?

Missing javadsl?

hseeberger commented:
Used "scala/akkadsl" convention. Added docs for JavaDSL. Please review again.

patriknw commented on sse/src/main/scala/akka/stream/alpakka/sse/javadsl/EventSource.scala:
```diff
@@ -0,0 +1,84 @@
+   * @return continuous source of server-sent events
+   */
+  def create(uri: Uri,
+             send: HttpRequest => CompletionStage[HttpResponse],
```
this will not work with Scala 2.11, so you should use `java.util.Function`

hseeberger commented:
@patriknw done

johanandren commented on docs/src/main/paradox/sse.md:
```diff
@@ -0,0 +1,46 @@
+
+## Usage
+
+### Sending messages to JMS server
```
copy pasta

hseeberger replied:
OMG ;-)

hseeberger commented:
Fixed the docs. Currently I really don't have any Java examples (although I do have a Java API). Maybe we can merge this and improve it going forward?

johanandren commented:
I'm ok with that.

hseeberger commented:
@patriknw are you also OK with that?

hseeberger commented:
@patriknw please review again.

patriknw commented:
Refs https://github.com/akka/alpakka/issues/47

patriknw commented:
Thanks @hseeberger 
I have created an issue for the addition of the Java examples: https://github.com/akka/alpakka/issues/173

hseeberger commented:
Thanks @patriknw
